example_plugins/keyboard/src/lib.rs

```diff
@@ -0,0 +1,197 @@
+                }) = event
+                {
+                    // Update key press information
+                    let is_pressed = state == ElementState::Pressed;
+
+                    if key == KeyCode::W {
+                        self.w_is_pressed = is_pressed;
+                    }
+
+                    if key == KeyCode::A {
+                        self.a_is_pressed = is_pressed;
+                    }
+
+                    if key == KeyCode::S {
+                        self.s_is_pressed = is_pressed;
+                    }
+
+                    if key == KeyCode::D {
+                        self.d_is_pressed = is_pressed;
+                    }
+                }
+            }
+        }
+
+        // Decide which way the cube should move based on keypresses
+        let mut move_vector = Vector3::zeros();
+        if self.w_is_pressed {
+            move_vector += Vector3::new(1., 0., 0.)
+        }
+        if self.a_is_pressed {
+            move_vector += Vector3::new(0., 0., -1.)
+        }
+        if self.s_is_pressed {
+            move_vector += Vector3::new(-1., 0., 0.)
+        }
+        if self.d_is_pressed {
+            move_vector += Vector3::new(0., 0., 1.)
+        }
+
+        // Send movement command to server
+        if move_vector != Vector3::zeros() {
+            let distance = move_vector.normalize() * frame_time.delta * 10.;
+
+            let command = MoveCommand { distance };
+
+            io.send(&command);
+        }
+    }
+}
+
+impl UserState for ServerState {
+    fn new(io: &mut EngineIo, sched: &mut EngineSchedule<Self>) -> Self {
+        // Define how the cube should be rendered
+
+        // Create one cube entity at the origin, and make it synchronize to clients
+        let cube_entity = io.create_entity();
+        io.add_component(cube_entity, &Transform::default());
+        io.add_component(
+            cube_entity,
+            &Render::new(CUBE_HANDLE).primitive(Primitive::Triangles),
+        );
+        io.add_component(cube_entity, &Synchronized);
+        io.add_component(cube_entity, &CubeFlag);
+
+        // Create the Update system, which interprets movement commands and updates the transform
+        // component on the object with CubeFlag
+        sched.add_system(
+            Self::update,
+            SystemDescriptor::new(Stage::Update)
+                .subscribe::<MoveCommand>()
+                .query::<CubeFlag>(Access::Write)
+                .query::<Transform>(Access::Write),
+        );
+
+        Self
+    }
+}
+
+impl ServerState {
+    fn update(&mut self, io: &mut EngineIo, query: &mut QueryResult) {
+        // Check for movement commands
+        if let Some(MoveCommand { distance }) = io.inbox_first() {
+            // Update each object accordingly
+            for key in query.iter() {
+                query.modify::<Transform>(key, |tf| {
+                    tf.pos += distance;
+                })
+            }
+        }
```

Comment:
This is more of a question on engine ergonomics, but how complicated is it to do what bevy does with Query<Transform> and be able to get the transform instead of the query modify? This seems like a very complicated API towards being able to modify items from the query.

I know it's most likely done out of ease of use but it seems like a very overly complicated system.

Reply:
You have to do fancy macro stuff to get what bevy has and ain't nobody got time for that. Also since we're serializing/deserializing it's a bit more complicated! (If you don't use `query::modify()`, then you won't pay the cost of deserialization. See also `query::read()` and `query::write()`!